Fixed bug 409 due to bad handling of NaN values in zoom linesearch

vroulet opened the pull request:
Fixed bug [409](https://github.com/google/jaxopt/issues/409), added corresponding test, enhanced debugging messages (for linesearch and solvers).

mblondel commented on tests/lbfgs_test.py:
```diff
@@ -358,6 +359,56 @@ def _run_lbfgs_solver(X, y, lmbd, n_iter):
+      lambda_xb = onp.exp(x@beta) / (1 + onp.exp(x@beta))
+      ll_i = y * onp.log(lambda_xb) + (1-y) * onp.log(1-lambda_xb)
+      ll = -onp.sum(ll_i)
+      return ll
```
remove

vroulet replied:
Which part do you want me to remove exactly? Having the comparison with scipy looks like a good safeguard. By the way I think we can remove the test test_first_stepsize_inf which did not reveal all issues.

mblondel replied:
I thought it's sufficient to just check that we don't get stuck at zero (our test suite is getting longer and longer to run). But you're right, maybe it's better to compare with scipy to be entirely sure.

vroulet replied:
Ok, I removed the previous test as it was redundant and less documented and kept the comparison to scipy.

mblondel commented on tests/lbfgs_test.py:
```diff
@@ -358,6 +359,56 @@ def _run_lbfgs_solver(X, y, lmbd, n_iter):
+    # using scipy
+    scipy_res = opt.minimize(fun=binary_logit_log_likelihood, 
+                             args=(onp.asarray(y),onp.asarray(x)), 
+                             x0 = (onp.asarray(beta_init)), method='BFGS')
```
remove

mblondel commented on jaxopt/_src/zoom_linesearch.py:
```diff
@@ -106,17 +106,17 @@ def _set_val(x, y):
 def _check_failure_status(fail_code):
   """Print failure reason according to fail value."""
-  if fail_code == 1:
+  if fail_code % 2 == 0:
```
Using bitwise flags would be more standard (I think).

vroulet commented:
I've also modified the way the linesearch would handle the case when the max_stepsize was reached. Before, once the maximal stepsize was reached the algorithm would keep on trying to increase it to find an interval for which the curvature error would be satisfied, while always staying to the max_stepsize. So there would be a waste of function/gradient evaluations. I fixed this by letting the linesearch stop if the max_stepsize is reached and no interval was found (meaning that the sufficient decrease condition is satisfied but not the curvature one and the latter would not be satisfied anyway). I documented it in the code.

mblondel commented on jaxopt/_src/zoom_linesearch.py:
```diff
@@ -138,7 +150,7 @@ class ZoomLineSearchState(NamedTuple):
   error: float
   done: bool
-  fail_code: int  # encode failure status, see _check_status
+  error_flag: int  # encode failure status, see _check_status
```
What's the motivation for changing the name? Since, other solvers use `fail_code`, this would create an inconsistency.

vroulet replied:
The NaN/Inf flag does not necessarily lead to a failure but was more of a warning. One could a priori fully remove it now too. Other than that, we can use fail_code. 
PS: Which other algorithms use fail_code? I could further harmonize if you want.

mblondel replied:
But in that case error != warning too. `fail_code` is only used in the zoom linesearch but `failed_linesearch` is used in many places so I would prefer to keep `fail_code` for consistency.

vroulet replied:
Ok! The changes have been made, I've squashed everything.
